lib/puppet/resource/status.rb

```diff
@@ -167,11 +171,13 @@ def initialize(resource)
         @time = Time.now
         @events = []
         @resource_type = resource.type.to_s.capitalize
+        @provider_used = resource.provider.class.to_s unless resource.provider.nil?
```

Comment:
`resource.provider.name.to_s` ?

Reply:
@joshcooper name isn't the thing we're looking for.

```
(byebug) resource.provider.name
"/my/file"
(byebug) resource.provider.class.to_s
"Puppet::Type::File::ProviderPosix"
(byebug) resource.provider.methods.sort
[:!, :!=, :!~, :<, :<=, :<=>, :==, :===, :=~, :>, :>=, :N_, :Nn_, :__id__, :__is_a__, :__metaclass__, :__send__, :adderrorcontext, :alert, :between?, :byebug, :class, :class_def, :clear, :clone, :command, :crit, :daemonize, :dclone, :debug, :debugger, :define_singleton_method, :devfail, :display, :dup, :emerg, :enum_for, :eql?, :equal?, :err, :error_context, :exceptwrap, :execfail, :execpipe, :execute, :expects, :extend, :fail, :feature?, :features, :fixtures, :freeze, :frozen?, :gem, :get, :get_posix_field, :gid, :gid2name, :group, :group=, :has_feature, :has_features, :hash, :idfield, :info, :inspect, :instance_eval, :instance_exec, :instance_of?, :instance_variable_defined?, :instance_variable_get, :instance_variable_set, :instance_variables, :is_a?, :itself, :kind_of?, :manages_symlinks?, :meta_def, :meta_eval, :meta_undef, :method, :method_exists?, :methodbyid, :methodbyname, :methods, :mocha, :mocha_inspect, :mode, :mode=, :my_fixture, :my_fixture_dir, :my_fixtures, :name, :name2gid, :name2uid, :nil?, :notice, :ns_, :object_id, :owner, :owner=, :pretty_inspect, :pretty_print, :pretty_print_cycle, :pretty_print_inspect, :pretty_print_instance_variables, :private_methods, :protected_methods, :psych_to_yaml, :public_method, :public_methods, :public_send, :remote_byebug, :remove_instance_variable, :reset_mocha, :resource, :resource=, :respond_to?, :s_, :satisfies?, :search_posix_field, :send, :set, :should, :should_not, :singleton_class, :singleton_method, :singleton_methods, :stubba_method, :stubba_object, :stubs, :taint, :tainted?, :tap, :to_enum, :to_json, :to_matcher, :to_pson, :to_s, :to_yaml, :to_yaml_properties, :trust, :uid, :uid2name, :unstub, :untaint, :untrust, :untrusted?, :warning, :with_verbose_disabled]
(byebug)
```

Reply:
Ah yeah, that's not right. I think you want:

```
(byebug) resource.provider.class.name
:posix
```

[Puppet's `genthing`](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/classgen.rb#L107-L124) metaprograms a `name` method on the class.